text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,508 @@
+  a construct library we're interested in (as determined by the `filePath`
+  where we found the exported class).
+
+The second case will give us slightly more information, although it might not
+necessarily be the most useful. Consider:
+
+```ts
+// Will be reported as 'Resource': class is not exported but anonymous class
+// extends 'Resource'.
+const bucket = Bucket.fromBucketName(this, 'ImportedBucket', 'my-bucket');
+
+// Will be reported as 'Construct' since it's a user-defined class and hence
+// not in scope.
+const myConstruct = new MyConstruct(this, 'MyConstruct', { ... });
+
+// On the other hand, the following might be interesting to report as 'Bucket'
+// (assuming the class 'extends Bucket').
+const richBucket = new RichBucket(this, 'RichBucket', { ... });
+```
+
+### Identifier components
+
+We obtain the construct identifier's components as follows:
+
+**CLASS NAME** We can get the simple class name by reading `obj.constructor.name`.
+
+**LIBRARY/VERSION** We can find the package a class is from by crawling up the
+directory tree from the filename we identified and looking for
+`package.json`, getting the `name` and `version` from that file.
+
+**SUBMODULE** Getting submodule names is more complex, as they don't have a
+runtime representation in NodeJS. We will have to derive the submodule from
+the file path and a helper file. The next section goes into more detail.
+
+### Submodule names
+
+Especially in monocdk and the upcoming v2, disambiguating identical class names
+by submodule is going to be essential. The question is how we will determine
+submodule names from file paths, keeping a couple of goals in mind:
+
+* In order to not proliferate identifier variations, it would be nice if the
+  submodule names would be the same names we use in the jsii build of CDK. There's
+  no strict need for them to be, but it just might ease future analysis if
+  we kept that consistency.
+* The mechanism should still work with libraries that don't have a submodule structure
+  at all, such as non-CDK team construct libraries vended by AWS like `@aws-solutions-constructs`.
+
+Some examples of what we want to derive:
+
+File name                                                              | Submodule
+-----------------------------------------------------------------------|----------
+.../node_modules/aws-cdk-lib/lib/aws-ecs/lib/cluster.js                | "aws_ecs"
+.../node_modules/aws-cdk-lib/lib/aws-eks/lib/cluster.js                | "aws_eks"
+.../node_modules/@aws-solutions-constructs/aws-lambda-sqs/lib/index.js | ""
+
+We could obtain submodule names by establishing conventions on monocdk's
+directory structure, using rules as "ignore `/lib/` and replace `-` with `_`
+to obtain the module name", for example. This reliance on convention seems
+brittle and unflexible. Instead, let's use a configuration file.
+
+One possible candidate is the jsii manifest. It contains the right module names
```

Comment:
I'm looking at the jsii manifest for `monocdk-experiment` and here's a snippet for eks Cluster and ecs Cluster -

```
      "name": "Cluster",
      "namespace": "aws_ecs",
      "properties": [
        {
          "docs": {
            "stability": "experimental",
            "summary": "The Amazon Resource Name (ARN) that identifies the cluster."
          },
          "immutable": true,
          "locationInModule": {
            "filename": "lib/aws-ecs/lib/cluster.ts",
            "line": 85
          },
          "name": "clusterArn",
          "overrides": "monocdk-experiment.aws_ecs.ICluster",
          "type": {
            "primitive": "string"
          }
        },
```

```
      "name": "Cluster",
      "namespace": "aws_eks",
      "properties": [
        {
          "docs": {
            "remarks": "This role also has `systems:master` permissions.",
            "stability": "experimental",
            "summary": "An IAM role with administrative permissions to create or update the cluster."
          },
          "immutable": true,
          "locationInModule": {
            "filename": "lib/aws-eks/lib/cluster.ts",
            "line": 685
          },
          "name": "adminRole",
          "type": {
            "fqn": "monocdk-experiment.aws_iam.Role"
          }
        },
```

It seems to me that from here we can derive the construct name (`name` property), module it is in (`namespace` property) and source file (`locationInModule` property).

Maybe we can update the JSII manifest to contain a `locationInModule` property for where the construct itself is, instead of individual properties.

What am I missing?

Reply:
> what if for some reason some sister team doesn't want to, but does want tracking?

I don't see a use case yet. This will be a nice-to-have, but we probably shouldn't index on this.

> To improve loading speed by a factor of 10 I'd want us to webpack the individual modules in the future (NodeJS takes a rather long time to load each and every individual source file), which means the jsii manifest will contain lib/aws-cdk/lib/cluster.ts but the runtime path we detect will look like aws-cdk/lib/aws-eks/index.js. Now we have to map that in addition.

Isn't your current algorithm also subject to the same problems if we change how we pack our output artifacts (webpack or otherwise). Things like - determining module name from file path, using the deepest export - come to mind.

Between this comment and https://github.com/aws/aws-cdk-rfcs/pull/254#discussion_r500150191, an option to generate the necessary metadata at compile time or even pack time (so that the path-to-module mapping is consistent) feels worth exploring. Don't you think? At least called out as an alternative.

Reply:
Just getting into the `cdk-metadata.json` section. Maybe this is what you're already doing that file.